ibis/backends/clickhouse/compiler/values.py

```diff
@@ -493,386 +409,290 @@ def _literal(op, **kw):
+    return f.ifNull(arg, 0)
 
 
-@translate_val.register(ops.DayOfWeekIndex)
-def _day_of_week_index(op, **kw):
-    arg = translate_val(op.arg, **kw)
-    weekdays = 7
-    offset = f"toDayOfWeek({arg})"
-    return f"((({offset} - 1) % {weekdays:d}) + {weekdays:d}) % {weekdays:d}"
+_NUM_WEEKDAYS = 7
 
 
-@translate_val.register(ops.FloorDivide)
-def _floor_divide(op, **kw):
-    new_op = ops.Floor(ops.Divide(op.left, op.right))
-    return translate_val(new_op, **kw)
+@translate_val.register(ops.DayOfWeekIndex)
+def _day_of_week_index(op, *, arg, **_):
+    weekdays = _NUM_WEEKDAYS
+    return (((f.toDayOfWeek(arg) - 1) % weekdays) + weekdays) % weekdays
 
 
-@translate_val.register(ops.ScalarParameter)
-def _scalar_param(op, params: Mapping[ops.Node, Any], **kw):
-    raw_value = params[op]
-    dtype = op.dtype
-    if isinstance(dtype, dt.Struct):
-        literal = ibis.struct(raw_value, type=dtype)
-    elif isinstance(dtype, dt.Map):
-        literal = ibis.map(raw_value, type=dtype)
-    else:
-        literal = ibis.literal(raw_value, type=dtype)
-    return translate_val(literal.op(), **kw)
+@translate_val.register(ops.FloorDivide)
+def _floor_divide(op, *, left, right, **_):
+    return f.floor(left / right)
 
 
 @translate_val.register(ops.StringContains)
-def _string_contains(op, **kw):
-    haystack = translate_val(op.haystack, **kw)
-    needle = translate_val(op.needle, **kw)
-    return f"locate({haystack}, {needle}) > 0"
+def _string_contains(op, haystack, needle, **_):
+    return f.locate(haystack, needle) > 0
 
 
 @translate_val.register(ops.InValues)
-def _in_values(op, **kw):
+def _in_values(op, *, value, options, **_):
     # TODO(kszucs): move this optimization to expression construction
     if not op.options:
-        return "FALSE"
-
-    value = translate_val(op.value, **kw)
-    if helpers.needs_parens(op.value):
-        value = helpers.parenthesize(value)
+        return FALSE
 
-    options = _sql(translate_val(op.options, **kw))
-
-    # we explicitly do NOT parenthesize the right side because it doesn't
-    # make sense to do so for Sequence operations
-    return f"{value} IN {options}"
+    return _parenthesize(op.value, value).isin(*options)
 
 
 @translate_val.register(ops.InColumn)
-def _in_column(op, **kw):
-    from ibis.backends.clickhouse.compiler import translate
-
-    value = translate_val(op.value, **kw)
-    if helpers.needs_parens(op.value):
-        value = helpers.parenthesize(value)
-
-    # this will fail to execute if there's a correlation, but it's too
-    # annoying to detect so we let it through to enable the
-    # uncorrelated use case (pandas-style `.isin`)
-    subquery = translate(op.options.to_expr().as_table().op(), {})
-    return f"{value} IN ({_sql(subquery)})"
+def _in_column(op, *, value, options, cache, aliases, **_):
+    base = an.find_first_base_table(op.options)
+    # TODO: cache is only used in this one place in the entire compiler; is it
```

Comment:
I would love to see it removed. Can you do `aliases.get(base)` instead?

Reply:
If cache would be removed then it could be removed from `node.map()` entirely.

Reply:
No.

Without this, the subquery doesn't get compiled because subquery is used here directly.

Reply:
In general we currently cannot know whether the subquery on the right hand side is already present somewhere in the query. Just because it's already been compiled doesn't mean it's been `SELECT`-ed from.

Reply:
Was able to remove this using a rewrite rule from `InColumn(left, right)` -> `InColumn(left, table_array_view(right))` in the compiler.